spinnaker/deploy_spinnaker/deploy_spinnaker.sh

```diff
@@ -0,0 +1,304 @@
+    else
+      dns_prefix="$REPLY"
+    fi
+  done
+}
+
+function get_ssh_key() {
+  local key_file_name="$1"
+  local key_files=($(ls $HOME/.ssh/*.pub))
+  local count=${#key_files[@]}
+  if [[ $count -gt 0 ]]; then
+    >&2 echo
+    >&2 echo "What ssh key would you like to use for your VM?"
+    for ((i = 0; i < $count; ++i)); do
+      if [ $i -lt 10 ]; then
+        >&2 echo "  [$i]  ${key_files[$i]}"
+      else
+        >&2 echo "  [$i] ${key_files[$i]}"
+      fi
+    done
+
+    local min=0
+    local max=$((count-1))
+    read -rp "===> Enter the number corresponding to the key file (leave blank to generate a new key): "
+    while [ -n "$REPLY" ] && (! [[ "$REPLY" =~ ^[0-9]+$ ]] || [ "$REPLY" -lt $min ] || [ "$REPLY" -gt $max ]); do
+      read -rp "===> Enter a valid integer between $min and $max (leave blank to generate a new key): "
+    done
+
+    if [ -n "$REPLY" ]; then
+      public_key_file="${key_files[$REPLY]}"
+      private_key_file="${public_key_file%????}"
+    fi
+  fi
+
+  if [ -z "$public_key_file" ]; then
+    private_key_file="$HOME/.ssh/${key_file_name}_rsa"
+    >&2 echo
+    >&2 echo "Generating ssh key..."
+    ssh-keygen -t rsa -b 2048 -N "" -f "$private_key_file" >/dev/null
+    public_key_file="${private_key_file}.pub"
+    >&2 echo "Generated key file '$public_key_file'."
+    >&2 echo "***IMPORTANT***: Add '-i $private_key_file' to any ssh command."
+  fi
+}
+
+function get_deploy_target() {
+  >&2 echo "This script will deploy an instance of Jenkins and Spinnaker on an Ubuntu 14.04 VM in Azure."
+  read -rp "===> Do you want Spinnaker to target a Kubernetes Cluster (enter 'k8s') or VM Scale Sets (enter 'vmss')?: "
+  deploy_target=$(echo "$REPLY" | awk '{print tolower($0)}')
+  while [ "$deploy_target" != "k8s" ] && [ "$deploy_target" != "vmss" ]; do
+    read -rp "===> Enter either 'vmss' or 'k8s': "
+    deploy_target=$(echo "$REPLY" | awk '{print tolower($0)}')
+  done
+}
+
+scenario_name=$(date +spinnaker-%Y-%m-%d-%H-%M-%S)
+
+get_deploy_target
+verify_az_cli
+set_subscription
+get_username
+get_dns_prefix
+create_resource_group "$scenario_name"
+create_service_principal "$scenario_name"
+
+if [ "$deploy_target" == "k8s" ]; then
+  get_ssh_key "$scenario_name"
+  key_file_note=", add '-i $private_key_file',"
+  template_uri="https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/301-jenkins-acr-spinnaker-k8s/azuredeploy.json"
+  parameters=$(cat <<EOF
+{
+  "adminUsername": {
+    "value": "$username"
+  },
+  "sshPublicKey": {
+    "value": "$(cat ${public_key_file})"
+  },
+  "devopsDnsPrefix": {
+    "value": "$dns_prefix"
+  },
+  "servicePrincipalAppId": {
+    "value": "$app_id"
+  },
+  "servicePrincipalAppKey": {
+    "value": "$app_key"
+  }
+}
+EOF
+)
+else
+  get_password
+  key_file_note=""
+  template_uri="https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/301-jenkins-aptly-spinnaker-vmss/azuredeploy.json"
```

Comment:
How we will make sure that the template parameters won't change and brake the script? I suggest adding an optional parameter for the template_uri and maybe we can add a tests in our CI